In [2]:
import pandas as pd
import numpy as np
import seaborn as sns
from pydataset import data

## Attendance Data



Read the data from the attendance table and calculate an attendance percentage for each student. One half day is worth 50% of a full day, and 10 tardies is equal to one absence.

You should end up with something like this:



name
Billy    0.5250
Jane     0.6875
John     0.9125
Sally    0.7625
Name: grade, dtype: float64


In [ ]:
attendance = pd.read_csv('untidy-data/attendance.csv')
attendance

In [ ]:
attendance.shape


In [ ]:
attendance = attendance.rename(columns={'Unnamed: 0': 'student'})
attendance

In [ ]:
attendance = attendance.melt(id_vars = 'student', var_name='date')
attendance

In [ ]:
#One half day is worth 50% of a full day, and 10 tardies is equal to one absence.
# tardie is one tenth so one tenth minus one is .9 
grades = {
    'P':1,
    'A':0,
    'H':0.5,
    'T':0.9,
}

In [ ]:
def attendance_grade(letter):
    if letter == 'P':
        return 1
    elif letter == 'A':
        return 0
    elif letter == 'H':
        return 0.5
    elif letter == 'T':
        return 0.9 

attendance_grade('T')    

In [ ]:

attendance['grade'] = att_melt.value.apply(attendance_grade)
attendance['grade']

In [ ]:
attendance.groupby('student').grade.mean()

## Coffee Levels

Read the coffee_levels table.
Transform the data so that each carafe is in it's own column.
Is this the best shape for the data?

In [ ]:
df = pd.read_csv('untidy-data/coffee_levels.csv')
df

In [ ]:
df = df.pivot_table(values='coffee_amount', index = 'hour', columns= 'coffee_carafe')
df

- Is this the best shape for the data?

yes its alot easier to read

## Cake Recipes



Read the cake_recipes table. This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.

In [4]:
df_cake = pd.read_csv('untidy-data/cake_recipes.csv')
df_cake #so the numbers inside the chart are the ammount of taistyness

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084
5           c:top  71.306308  82.795477  92.098049  53.960273
6        d:bottom  52.799753  58.670419  51.747686  56.183110
7           d:top  96.873178  76.101363  59.571620  50.971626

Tidy the data as necessary.

In [5]:
cake_change = df_cake['recipe:position'].str.split(':',expand = True)
cake_change

0       1
0  a  bottom
1  a     top
2  b  bottom
3  b     top
4  c  bottom
5  c     top
6  d  bottom
7  d     top

In [7]:
cake_change.columns = ['recipe', 'position']

cake_change

recipe position
0      a   bottom
1      a      top
2      b   bottom
3      b      top
4      c   bottom
5      c      top
6      d   bottom
7      d      top

In [8]:
df_cake = pd.concat([df_cake,cake_change], axis=1).drop(columns ='recipe:position')
df_cake

225        250        275        300 recipe position
0  61.738655  53.912627  74.414730  98.786784      a   bottom
1  51.709751  52.009735  68.576858  50.228470      a      top
2  57.095320  61.904369  61.196980  99.248541      b   bottom
3  82.455004  95.224151  98.594881  58.169349      b      top
4  96.470207  52.001358  92.893227  65.473084      c   bottom
5  71.306308  82.795477  92.098049  53.960273      c      top
6  52.799753  58.670419  51.747686  56.183110      d   bottom
7  96.873178  76.101363  59.571620  50.971626      d      top

In [9]:
df_cake = df_cake.melt(id_vars = ['recipe' , 'position'],var_name= 'temp', value_name= 'tastiness')
df_cake

recipe position temp  tastiness
0       a   bottom  225  61.738655
1       a      top  225  51.709751
2       b   bottom  225  57.095320
3       b      top  225  82.455004
4       c   bottom  225  96.470207
5       c      top  225  71.306308
6       d   bottom  225  52.799753
7       d      top  225  96.873178
8       a   bottom  250  53.912627
9       a      top  250  52.009735
10      b   bottom  250  61.904369
11      b      top  250  95.224151
12      c   bottom  250  52.001358
13      c      top  250  82.795477
14      d   bottom  250  58.670419
15      d      top  250  76.101363
16      a   bottom  275  74.414730
17      a      top  275  68.576858
18      b   bottom  275  61.196980
19      b      top  275  98.594881
20      c   bottom  275  92.893227
21      c      top  275  92.098049
22      d   bottom  275  51.747686
23      d      top  275  59.571620
24      a   bottom  300  98.786784
25      a      top  300  50.228470
26      b   bottom  300  99.248541
27      b      top  300  58.169349
28      c   bottom  300  65.473084
29      c      top  300  53.960273
30      d   bottom  300  56.183110
31      d      top  300  50.971626

Which recipe, on average, is the best? recipe b

In [11]:
df_cake.groupby('recipe').tastiness.mean()

recipe
a    63.922201
b    76.736074
c    75.874748
d    62.864844
Name: tastiness, dtype: float64

Which oven temperature, on average, produces the best results? 275

In [12]:
df_cake.groupby('temp').tastiness.mean()

temp
225    71.306022
250    66.577437
275    74.886754
300    66.627655
Name: tastiness, dtype: float64

Which combination of recipe, rack position, and temperature gives the best result? recipe b, bottom rack, 300 degrees

In [13]:
df_cake.sort_values(by = 'tastiness').tail()

recipe position temp  tastiness
4       c   bottom  225  96.470207
7       d      top  225  96.873178
19      b      top  275  98.594881
24      a   bottom  300  98.786784
26      b   bottom  300  99.248541

## Bonus: explore the other tables in the tidy_data database and reshape them as necessary so that they are in a tidy format.

In [20]:
sales = pd.read_csv('untidy-data/sales.csv')

In [21]:
sales_melt = sales.melt(id_vars='Product', var_name='year_and_measure')
sales_melt.head()


Product year_and_measure  value
0       A       2016 Sales    673
1       B       2016 Sales    259
2       C       2016 Sales    644
3       D       2016 Sales    508
4       A         2016 PPU      5

In [25]:
year_and_measure_df = sales_melt.year_and_measure.str.split(' ',expand=True)
year_and_measure_df.columns= ['year','measure']
year_and_measure_df

year measure
0   2016   Sales
1   2016   Sales
2   2016   Sales
3   2016   Sales
4   2016     PPU
5   2016     PPU
6   2016     PPU
7   2016     PPU
8   2017   Sales
9   2017   Sales
10  2017   Sales
11  2017   Sales
12  2017     PPU
13  2017     PPU
14  2017     PPU
15  2017     PPU
16  2018   Sales
17  2018   Sales
18  2018   Sales
19  2018   Sales
20  2018     PPU
21  2018     PPU
22  2018     PPU
23  2018     PPU

In [29]:
sales2 = pd.concat([sales_melt, year_and_measure_df], axis=1).drop(columns='year_and_measure')
sales2

Product  value  year measure
0        A    673  2016   Sales
1        B    259  2016   Sales
2        C    644  2016   Sales
3        D    508  2016   Sales
4        A      5  2016     PPU
5        B      3  2016     PPU
6        C      3  2016     PPU
7        D      9  2016     PPU
8        A    231  2017   Sales
9        B    748  2017   Sales
10       C    863  2017   Sales
11       D    356  2017   Sales
12       A      7  2017     PPU
13       B      5  2017     PPU
14       C      5  2017     PPU
15       D     11  2017     PPU
16       A    173  2018   Sales
17       B    186  2018   Sales
18       C    632  2018   Sales
19       D    347  2018   Sales
20       A      9  2018     PPU
21       B      8  2018     PPU
22       C      5  2018     PPU
23       D     14  2018     PPU

In [31]:
sales_tidy = sales2.pivot_table(index=['Product', 'year'], columns='measure', values='value')
sales_tidy


measure       PPU  Sales
Product year            
A       2016    5    673
        2017    7    231
        2018    9    173
B       2016    3    259
        2017    5    748
        2018    8    186
C       2016    3    644
        2017    5    863
        2018    5    632
D       2016    9    508
        2017   11    356
        2018   14    347

In [32]:
sales_tidy.columns.name = ''
sales_tidy.reset_index(inplace=True)
sales_tidy


Product  year  PPU  Sales
0        A  2016    5    673
1        A  2017    7    231
2        A  2018    9    173
3        B  2016    3    259
4        B  2017    5    748
5        B  2018    8    186
6        C  2016    3    644
7        C  2017    5    863
8        C  2018    5    632
9        D  2016    9    508
10       D  2017   11    356
11       D  2018   14    347